### DISCUSSION TITLE
Use numerical functions instead of symbolic

### DISCUSSION THREAD

**Discussion by agentkashtan** [casadi/casadi#3795] [Q&A] [answered]
Hi everyone,

I'm new to Casadi and trying to implement mpc for a 7 DOF robot. I need to compute the next state given the current and inputs i.e. $$x_{k+1} = x_{k} + f(x_{k}, u_{k}) \cdot dt$$. The issue I have is that symbolic expressions are huge(probably will take forever to optimize) for matrices I need to compute dynamics. OTOH, I have an external library that can numerically compute those matrices for the current state. I wonder if its somehow possible to use those numerical values and still be able to run the optimizer.

I would like to ask for any suggestions on how to achieve this.

I've added a code snippet below for clarity(I don't know how to handle M_symbolic, C_symbolic part). Thank you in advance, sorry if the question is not clear enough. 
```
def numerical_M(q):
    M = ... external lib call, depends on numerical q...
    return M

def numerical_C(q, q_dot):
    C = ... external lib call, depends on numerical q, q_dot...
    return C

N = 20  
T = 2.0  
dt = T / N 

X = ca.MX.sym('X', 14, N + 1)  # 7 for q and 7 for q_dot
U = ca.MX.sym('U', 7, N)  # Controls over the horizon

# Define the initial state as a parameter
X0 = ca.MX.sym('X0', 14)


objective = 0
g = []  # Constraint vector

for k in range(N):
    qk = X[:7, k]
    q_dotk = X[7:, k]
    uk = U[:, k]
    
    
    M_symbolic = somehow  get M
    C_symbolic = somehow  get C
    
    q_ddotk = ca.mldivide(M_symbolic, uk - C_symbolic)
    
    q_dot_next = q_dotk + dt * q_ddotk
    q_next = qk + dt * q_dot_next
    
    x_next = ca.vertcat(q_next, q_dot_next)
    
    g.append(X[:, k+1] - x_next)
    
    objective += ....
```

**Comment by charleskhazoom** [comment]
So you already have expressions for M_symbolic and C_symbolic ? Did you try to solve your problem with these "huge expressions" ? A 7-DoF robot with N=20 seems like something that Casadi should handle without problem.

**Reply by agentkashtan** [reply]
I haven't tried actual symbolic expressions for M and C, because there is a similar question in the discussions and the author says that it takes too long. But I'll try with my symbolic expressions, maybe there were some mistakes in that post. 
Thank you.

**Comment by tmmsartor** [comment, marked as answer]
If you have large symbolic expression in `M_symbolic`, consider there may be way to simplify it and/or speed it up, i.e. having a hierarchy of Functions, using [foor-loop equivalent](http://docs.casadi.org/#for-loop-equivalents) or  code generating the functions needed by the solver.


If that is not an option, to use the output of `numerical_C(q, q_dot)` you can wrap it it  casadi `Callback`, and then use in place of `M_symbolic` in you expression, see example [here](https://github.com/casadi/casadi/blob/develop/docs/examples/python/callback.py).
Be aware that in order for the optimization solver to work you would most likely need to provide also some derivatives, i.e. d(numerical_C(q, q_dot))/dq, d(numerical_C(q, q_dot))/dq_dot, if your external library do not provide those you may resort to use finite differences, those can also be defined inside the given Callback method, i.e. `get_reverse`

**Reply by agentkashtan** [reply]
Thank you. So if I want to optimize over a function, then for Casadi to work it's enough to define a "wrapper" of my numerical representation of the actual function that returns value and gradient numerically given numeric inputs?

**Reply by tmmsartor** [reply]
The derivatives you need to provide depend on the solver your are using. If are using a second order solver, i.e. `ipopt` with full hessian, you need also second order derivatives. If you are using an hessian approximation, like when passing the option `"hessian_approximation":"limited-memory"` to ipopt you should need only the jacobian.
